xilem/src/view/worker.rs

```diff
@@ -33,13 +35,47 @@ pub fn worker<F, H, M, S, V, State, Action, Fut>(
     init_future: F,
     store_sender: S,
     on_response: H,
-) -> Worker<F, H, M, S, V>
+) -> Worker<F, H, M, impl Fn(&mut State, &mut Dummy, UnboundedSender<V>) + 'static, V, Dummy>
 where
     F: Fn(MessageProxy<M>, UnboundedReceiver<V>) -> Fut,
     Fut: Future<Output = ()> + Send + 'static,
-    S: Fn(&mut State, UnboundedSender<V>),
+    S: Fn(&mut State, UnboundedSender<V>) + 'static,
     H: Fn(&mut State, M) -> Action + 'static,
     M: AnyDebug + Send + 'static,
+{
+    const {
+        assert!(
+            size_of::<F>() == 0,
+            "`worker` will not be ran again when its captured variables are updated.\n\
+            To ignore this warning, use `worker_raw`."
+        );
+    };
+    Worker {
+        init_future,
+        store_sender: move |state: &mut State, _dummy: &mut Dummy, sender: UnboundedSender<V>| {
+            store_sender(state, sender);
+        },
+        on_response,
+        message: PhantomData,
+    }
+}
+
+/// A version of [`worker`] which can store its message sender in the environment.
+///
+/// This is an interim solution until we work out nicer
+/// interactions with environment values in callbacks and the like.
+pub fn env_worker<F, H, M, S, V, Res, State, Action, Fut>(
+    init_future: F,
+    store_sender: S,
+    on_response: H,
+) -> Worker<F, H, M, S, V, Res>
+where
+    F: Fn(MessageProxy<M>, UnboundedReceiver<V>) -> Fut,
+    Fut: Future<Output = ()> + Send + 'static,
+    S: Fn(&mut State, &mut Res, UnboundedSender<V>),
+    H: Fn(&mut State, M) -> Action + 'static,
```

Comment:
I'm wondering whether it would be more consistent to also have `&mut Res` in this callback?

Reply:
Probably, but I think that would make more sense as part of a generalised query system.

The reason I only added it to `store_sender` here was because the general pattern I want to use (for now) is `on_action_with_context` and returning an explicit action. But that doesn't work for `store_sender`, because it's not actually a true action.